docs/source/guides/schema-design.md

```diff
@@ -0,0 +1,290 @@
+
+Furthermore, if we need to return fields which are only provided by either `TextBook`s or `ColoringBook`s (not both) we can request fragments from the abstract types in the query. Those fragments will be filled in only as appropriate; in the case of the example, only coloring books will be returned with `colors`, and only textbooks will have `classes`:
+
+```graphql
+query GetBooks {
+  schoolBooks {
+    title
+    ... on TextBook {
+      classes {
+        name
+      }
+    }
+    ... on ColoringBook {
+      colors {
+        name
+      }
+    }
+  }
+}
+```
+
+<h2 id="mutations">Designing mutations</h2>
+
+Mutations are one of the core types in GraphQL. Just like you can make a query to fetch information, you can make a mutation to update or change information. Unlike REST, GraphQL mutations actually are executed in two parts: the mutation itself, and a subsequent query, which can return any kind of data that you normally could query for. A mutation definition for updating the age of a `User` could look like this:
+
+```graphql
+type Mutation {
+  updateUserAge(id: ID!, age: Int!): User
+}
+
+type User {
+  id: ID!
+  name: String!
+  age: Int!
+}
+```
+
+With that definition, you could then make the following mutation:
+
+```graphql
+mutation updateMyUser {
+  updateUserAge(id: 1, age: 25){
+    id
+    age
+    name
+  }
+}
+```
+
+With a response looking something like:
+
+```json
+{
+  "data": {
+    "updateUserAge": {
+      "id": "1",
+      "age": "25",
+      "name": "jane doe"
+    }
+  }
+}
+```
+
+The first thing to note is that it’s most common to return the thing you’re updating from a mutation. In our example, we were updating a `User` record, so we returned that updated user. There’s nothing _enforcing_ this practice, but it’s highly recommended, because it’s often needed to have an updated user to let the clients update any local cache with a new instance of that `User`. For this same reason, it’s useful design mutations to update only one entity. If you wanted to update two unrelated entities, it’s recommended to in separate mutations. For more details on how to handle errors and warnings in mutations, see the [mutation responses](#mutation-responses) section below.
+
+But what if we wanted to update more than just a single or couple attributes on a user? Passing each thing we need as a single argument would get tedious. Especially if multiple mutations used similar fields. For this, we can use input types, which are explained in the next section.
+
+<h3 id="mutation-input-types">Input types</h3>
+
+Input types are a special type in GraphQL which are defined as arguments to queries and, more commonly, mutations.  They can be thought of as object types for arguments, in addition to the other scalar types.  Input types are especially useful when multiple mutations require similar information; for example, when creating a user and updating a user require the same fields, like `age` and `name`.
```

Comment:
I think object types are actually most useful when a single mutation requires many arguments, and I'm not actually sure that reusing input types across mutations is a good idea. For example, in the `updateUser` case, you probably want most of the fields to be nullable, in case you want to update just some of them, but in the `createUser` case you might want to enforce that they are all present.

Even if that isn't the case at first, you might eventually need to change your arguments such that they do have different input requirements, and changing the input type of an argument is always a breaking change for existing clients (even if the fields are compatible).

So I'd suggest we change this recommendation to say:

1. We suggest input types for mutations which have a complex set of arguments
2. We suggest not sharing input types across multiple fields